nni/compression/pytorch/speedup/compress_modules.py

```diff
@@ -3,196 +3,381 @@
     return module
 
 
-def replace_linear(linear, mask):
+def replace_linear(linear, auto_infer):
     """
     Parameters
     ----------
     linear : torch.nn.Linear
         The linear module to be replace
-    mask : ModuleMasks
-        The masks of this module
+    auto_infer : AutoMaskInference
+        The auto mask inference object that contains the input,
+        parameter and output masks.
 
     Returns
     -------
     torch.nn.Linear
         The new linear module
     """
-    assert isinstance(mask, ModuleMasks)
-    assert mask.input_mask is not None
-    assert mask.output_mask is None
-    assert not mask.param_masks
-    index = mask.input_mask.mask_index[-1]
-    in_features = index.size()[0]
-    _logger.debug("replace linear with new in_features: %d", in_features)
-    new_linear = torch.nn.Linear(in_features=in_features,
-                                 out_features=linear.out_features,
-                                 bias=linear.bias is not None)
+    NEED_FOLD_BIAS = auto_infer.fold_bias
+    assert isinstance(linear, nn.Linear)
+    assert len(auto_infer.in_masks) == 1
+    assert isinstance(auto_infer.output_mask, torch.Tensor)
+    in_mask = auto_infer.in_masks[0]
+    # only need the first batch of the constant
+    in_constant = auto_infer.in_constants[0][:1]
+    output_mask = auto_infer.output_mask
+    weight_mask = auto_infer.weight_mask['weight']
+    # N C K
+    pruned_in, remained_in = convert_to_coarse_mask(in_mask, 1)
+    pruned_out, remained_out = convert_to_coarse_mask(output_mask, 1)
+    n_remained_in = weight_mask.size(1) - pruned_in.size(0)
+    n_remained_out = weight_mask.size(0) - pruned_out.size(0)
+    remained_in, remained_out = remained_in.to(
+        linear.weight.device), remained_out.to(linear.weight.device)
+    _logger.info("replace linear with new in_features: %d, out_features: %d",
+                 n_remained_in, n_remained_out)
+    need_bias = False
+    if linear.bias is not None or (NEED_FOLD_BIAS and torch.sum(in_constant > 0)):
+        need_bias = True
+    new_linear = torch.nn.Linear(in_features=n_remained_in,
+                                 out_features=n_remained_out,
+                                 bias=need_bias)
     new_linear.to(linear.weight.device)
-    new_linear.weight.data = torch.index_select(
-        linear.weight.data, -1, index.to(linear.weight.device))
-    if linear.bias is not None:
-        new_linear.bias.data.copy_(linear.bias.data)
+    # Copy the remained weight from the original module
+    with torch.no_grad():
+        tmp_weight_data = torch.index_select(
+            linear.weight.data, 0, remained_out)
+        new_linear.weight.data = torch.index_select(
+            tmp_weight_data, 1, remained_in)
+
+        if linear.bias is not None:
+            new_linear.bias.data = torch.index_select(
+                linear.bias.data, 0, remained_out)
+
+    if NEED_FOLD_BIAS and torch.sum(in_constant) > 0:
```

Comment:
let's discuss more about bias folding. can this type of folding be easily generalized to other operators?